ompi/runtime/ompi_spc.c

```diff
@@ -0,0 +1,447 @@
+ * of the counter values to stdout.
+ */
+void ompi_spc_dump(void)
+{
+    int i, j, world_size, offset;
+    long long *recv_buffer = NULL, *send_buffer;
+
+    ompi_communicator_t *comm = &ompi_mpi_comm_world.comm;
+    int rank = ompi_comm_rank(comm);
+    world_size = ompi_comm_size(comm);
+
+    /* Convert from cycles to usecs before sending */
+    for(i = 0; i < OMPI_SPC_NUM_COUNTERS; i++) {
+        if( IS_SPC_BIT_SET(ompi_spc_timer_event, i) ) {
+            SPC_CYCLES_TO_USECS(&ompi_spc_events[i].value);
+        }
+    }
+
+    /* Aggregate all of the information on rank 0 using MPI_Gather on MPI_COMM_WORLD */
+    send_buffer = (long long*)malloc(OMPI_SPC_NUM_COUNTERS * sizeof(long long));
+    if(send_buffer == NULL) {
+        opal_show_help("help-mpi-runtime.txt", "spc: failed to allocate memory", true);
+        return;
+    }
+    for(i = 0; i < OMPI_SPC_NUM_COUNTERS; i++) {
+        send_buffer[i] = (long long)ompi_spc_events[i].value;
+    }
+    if( 0 == rank ) {
+        recv_buffer = (long long*)malloc(world_size * OMPI_SPC_NUM_COUNTERS * sizeof(long long));
+        if(recv_buffer == NULL) {
+            opal_show_help("help-mpi-runtime.txt", "spc: failed to allocate memory", true);
+            return;
+        }
+    }
+    (void)comm->c_coll->coll_gather(send_buffer, OMPI_SPC_NUM_COUNTERS, MPI_LONG_LONG,
+                                    recv_buffer, OMPI_SPC_NUM_COUNTERS, MPI_LONG_LONG,
+                                    0, comm,
+                                    comm->c_coll->coll_gather_module);
+
+    /* Once rank 0 has all of the information, print the aggregated counter values for each rank in order */
+    if(rank == 0) {
+        opal_output(0, "Open MPI Software-based Performance Counters:\n");
+        offset = 0; /* Offset into the recv_buffer for each rank */
+        for(j = 0; j < world_size; j++) {
+            opal_output(0, "MPI_COMM_WORLD Rank %d:\n", j);
+            for(i = 0; i < OMPI_SPC_NUM_COUNTERS; i++) {
+                /* If this is a timer-based counter, we need to covert from cycles to usecs */
+                if( 0 == recv_buffer[offset+i] ) {
+                    continue;
+                }
+                opal_output(0, "%s -> %lld\n", ompi_spc_events[i].name, recv_buffer[offset+i]);
+            }
+            opal_output(0, "\n");
+            offset += OMPI_SPC_NUM_COUNTERS;
+        }
+        printf("###########################################################################\n");
+        printf("NOTE: Any counters not shown here were either disabled or had a value of 0.\n");
+        printf("###########################################################################\n");
+
+        free(recv_buffer);
+    }
+    free(send_buffer);
+
+    comm->c_coll->coll_barrier(comm, comm->c_coll->coll_barrier_module);
+}
+
+/* Frees any dynamically alocated OMPI SPC data structures */
+void ompi_spc_fini(void)
+{
+#if SPC_ENABLE == 1
+    ompi_spc_dump();
```

Comment:
It looks like I forgot to add the check for whether dumping is enabled.  I'll add this when I get home this evening.